buffer/test/com/treode/buffer/PagedBufferSpec.scala

```diff
@@ -48,33 +48,52 @@ class PagedBufferSpec extends FlatSpec {
   capacity (64, 2)
   capacity (65, 3)
 
-  def discard (nbytes: Int, npages: Int) {
-    it should (s"discard the right pages for nbytes=$nbytes") in {
+  def discard (nbytes: Int, npages: Int, wbytes: Int, rbytes: Int, dbytes: Int) {
```

Comment:
Ah, good catch. In that case, I think the test parameters can be:

    discard (nbytes, rbytes, wbytes, npages)
    // nbytes -> number of bytes to discard
    // rbytes -> set readPos, must be >= nbytes
    // wbytes -> set writePos, must be >= rbytes
    // npages -> expected number of discarded pages

Note, I'm specifically listing the "control" parameters first, and the expected values after.

We don't need to pass in the expected number of discarded bytes, because we can compute it.

    val discarded = npages * pageSize

Instead of making a second buffer to set readPos and writePos, you can just set them on the first buffer. You can assert the readPos and writePos are modified as expected on the first buffer.

Since we are varying readPos and writePos now, we should have a few tests that do so while holding nbytes constant.  For example, where the test previously listed:

    discard (32, 1) // nbytes, npages

We now need a handful of cases:

    discard (32, 1, 32, 53)
    discard (32, 1, 42, 53)
    discard (32, 1, 64, 64)
    // etc

When you do this, Scalatest is going to complain about tests with duplicate names. You need to include all control parameters in the test name:

    it should (s"discard the right pages for nbytes=$nbytes")

would become

    it should (s"discard the right pages for nbytes=$nbytes, rbytes=$rbytes, wbytes=$wbytes")

I originally listed discard tests without much comment about what they're doing. Just because I did it doesn't mean it's good. I think it would help to add comments like:

  // No-op discard, because nbytes < pageSize
  discard (0, 0, 0, 0, 0) // all values aligned at page start
  discard (7, 0, 6, 5, 5) // nbytes aligned at page start
  discard (31, 0, 32, 31, 0) // nbytes aligned at page end

The comments would highlight which control values are aligned, at let us see at a glance that we enumerated all possible combinations of some aligned and some not.

I see you did things like:

    assertResult (rbytes - newreadpos) (discarded)

When this assertion fails, it would give a message like:

    "Expected ${rbytes - newreadpos}, but found $discarded."

This error message has the expected and actual backwards, so it could be misleading. Swap the parameters so that the first one to `assertResult` is the expected value.